packages/flutter_tools/lib/src/version.dart

```diff
@@ -96,11 +153,141 @@ class FlutterVersion {
+  ║ $warning ║
+  ║                                                                            ║
+  ║ To update to the latest version, run flutter upgrade.                      ║
+  ╚════════════════════════════════════════════════════════════════════════════╝
+''';
+  }
+
+  /// Gets the release date of the latest available Flutter version.
+  ///
+  /// This method sends a server request if it's been more than
+  /// [kCheckAgeConsideredUpToDate] since the last version check.
+  ///
+  /// Returns `null` if the cached version is out-of-date or missing, and we are
+  /// unable to reach the server to get the latest version.
+  Future<DateTime> _getLatestAvailableFlutterVersion() async {
+    Cache.checkLockAcquired();
+    const JsonEncoder kPrettyJsonEncoder = const JsonEncoder.withIndent('  ');
+    final String versionCheckStamp = Cache.instance.getStampFor(kFlutterVersionCheckStampFile);
+
+    if (versionCheckStamp != null) {
+      final Map<String, String> data = JSON.decode(versionCheckStamp);
+      final DateTime lastTimeVersionWasChecked = DateTime.parse(data['lastTimeVersionWasChecked']);
+      final Duration timeSinceLastCheck = clock.now().difference(lastTimeVersionWasChecked);
+
+      // Don't ping the server too often. Return cached value if it's fresh.
+      if (timeSinceLastCheck < kCheckAgeConsideredUpToDate) {
+        return DateTime.parse(data['lastKnownRemoteVersion']);
+      }
+    }
+
+    // Cache is empty or it's been a while since the last server ping. Ping the server.
+    try {
+      final DateTime remoteFrameworkCommitDate = DateTime.parse(await FlutterVersion.fetchRemoteFrameworkCommitDate());
+      Cache.instance.setStampFor(kFlutterVersionCheckStampFile, kPrettyJsonEncoder.convert(<String, String>{
+        'lastTimeVersionWasChecked': '${clock.now()}',
+        'lastKnownRemoteVersion': '$remoteFrameworkCommitDate',
+      }));
+      return remoteFrameworkCommitDate;
+    } on ToolExit catch(error) {
+      // This happens when any of the git commands fails, which can happen when
+      // there's no Internet connectivity. Remote version check is best effort
+      // only. We do not prevent the command from running when it fails.
+      printTrace('Failed to check Flutter version in the remote repository: ${error.message}');
+      return null;
+    }
+  }
 }
 
-String _runSync(List<String> command, String cwd) {
+/// Runs [command] and returns the standard output as a string.
+///
+/// If [lenient] is `true` and the command fails, returns an empty string.
+/// Otherwise, throws a [ToolExit] exception.
+String _runSync(List<String> command, String cwd, {bool lenient: true}) {
   final ProcessResult results = processManager.runSync(command, workingDirectory: cwd);
-  return results.exitCode == 0 ? results.stdout.trim() : '';
+
+  if (results.exitCode == 0)
+    return results.stdout.trim();
+
+  if (!lenient) {
+    throwToolExit(
```

Comment:
ToolExit seems like a violent exception to throw here. It's also confusing because that has the semantics of "the tool will now exit" but then you catch it, which defeats those semantics. I suggest having a separate exception, e.g. "VersionCheckError" or some such, specifically for this purpose.